JDBC/OAuthSampleApp/src/main/java/OAuthSampleApp.java

```diff
@@ -70,14 +147,128 @@ private static void printResults(ResultSet rs) throws SQLException {
+	            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+	            connection.setRequestProperty("Content-Length", Integer.toString(postDataLength));
+	            connection.setRequestProperty("Accept", "application/json");
+				connection.getOutputStream().write(postData);
+	           	BufferedInputStream in = new BufferedInputStream(connection.getInputStream());
+	    	    ByteArrayOutputStream buf = new ByteArrayOutputStream();
+	            for (int result = in.read(); result != -1; result = in.read()) {
+	               	buf.write((byte) result);
+		        }
+	    	    String res = buf.toString("UTF-8");
+	        	JsonElement jElement = new JsonParser().parse(res);
+	            JsonObject jObject = jElement.getAsJsonObject();
+	            
+				// set Tokens as System Properties - new values to access_token and refresh_token
+				System.setProperty(OAUTH_ACCESS_TOKEN_VAR_STRING, jObject.get("access_token").getAsString());
+				System.setProperty(OAUTH_REFRESH_TOKEN_VAR_STRING, jObject.get("refresh_token").getAsString());
+
+	        } catch (Exception e) {
+	            String res = "";
+	            try {
+	                BufferedInputStream in = new BufferedInputStream(connection.getErrorStream());
+	                ByteArrayOutputStream buf = new ByteArrayOutputStream();
+	                for (int result = in.read(); result != -1; result = in.read()) {
+	                    buf.write((byte) result);
+	                }
+	                res = buf.toString("UTF-8");
+	            } catch (Exception ex) {
+	                //Improper setup. Passes in SF, fails in Devjail. Skip when this happens, but print the error.
+	                ex.printStackTrace();
+	                System.out.println("Error when fetching token endpoints. Skipping test.");
+	            }
+				throw e;
+	        } finally {
+	            connection.disconnect();
+	        }
+    	}catch (IOException unreportedex) {
+    		unreportedex.printStackTrace();
+        }
+    }
+    
+    // if no access token is set, obtains and sets first access and refresh tokens
+    // uses the password grant flow
+    private void EnsureAccessToken() throws Exception {
+		System.out.println("EnsureAccessToken");
+		try {
+			String accessToken = System.getenv(OAUTH_ACCESS_TOKEN_VAR_STRING);
+			if(null == accessToken || accessToken.isEmpty()) {
+				// Obtain first access token
+				System.out.println("Access token not found. Obtaining first access token from IDP.");
+				GetTokensByPasswordGrant();
+			}
+		}catch(Exception e)
+		{
+			throw e;
+		}
+	}
+    
     @Override
     public Integer call() throws Exception {
         try {
-            Connection conn = connectToDB(host, port, dbName, accessToken, refreshToken, clientSecret);
-            ResultSet rs = executeQuery(conn);
-            printResults(rs);
-            conn.close();
+       		//SetUpDbForOAuth();
+        	EnsureAccessToken();
+        	ConnectToDatabase();
```

Comment:
Here you are setting the access and refresh tokens and then trying to connect. If that works, then you call "GetTokensByRefreshGrant" which sets the access and refresh tokens. I don't understand the logic.

Reply:
Removed this GetTokensByRefreshGrant call and added to connectToDatabase() method as per ADO.Net code flow.